src/chain/auth.sw

```diff
@@ -1,41 +1,160 @@
+        // get data offest by 1 word
+        let data_ptr = asm(buffer, ptr: input_ptr, data_ptr) {
+            move buffer sp;
+            cfei i8;
+            addi data_ptr input_ptr i8;
+            mcpi buffer data_ptr i8;
+            buffer: u8
+        };
+
+        let owner_addr = ~Address::from(asm(buffer, ptr: data_ptr, owner_ptr) {
+            move buffer sp;
+            cfei i8;
+            addi owner_ptr data_ptr i16;
+            mcpi buffer owner_ptr i32;
+            buffer: b256
         });
-        Result::Ok(Sender::Id(id))
+
+        Result::Ok(Sender::Address(owner_addr))
+
+}
+
+/// Get the owner of the inputs(of type `InputCoin`) to a TransactionScript,
+/// if they all share the same owner.
+pub fn get_coins_owner() -> Result<Sender> {
+    let zero_addr = ~Address::from(0x0000000000000000000000000000000000000000000000000000000000000000);
+    let target_input_type = 0u8;
+    let inputs_count = get_inputs_count();
+
+    let mut candidate = ~Address::from(0x0000000000000000000000000000000000000000000000000000000000000000);
+    let mut input_owner = ~Address::from(0x0000000000000000000000000000000000000000000000000000000000000000);
+    let mut i = 0u8;
+    let mut input_pointer: u32 = 0u32;
+    let mut input_type: u8 = 0u8;
+
+    while i < inputs_count {
+        input_pointer = get_input_pointer(i);
+        input_type = get_input_type(input_pointer);
+        if input_type != target_input_type {
+            // increment the counter and continue looping
+            i = i + 1;
+        } else {
+            // type == InputCoin
+            if candidate == zero_addr {
+                // this is the first input seen of the correct type
+                input_owner = get_input_owner(input_pointer);
+                candidate = input_owner;
+            } else {
+                // compare current coin owner to candidate
+                if input_owner == candidate {
+                    // owners are a match, continue looping
+                    i = i + 1;
+                } else {
+                    // owners don't match. Break and return Err
+                    i = inputs_count;
+                    Result::Err(AuthError::ContextError)
+               };
+            };
+        };
+    }
+    Result::Ok(Sender::Address(candidate))
+}
+
+/// Get a pointer to an input given the index of the input.
+fn get_input_pointer(n: u8) -> u32 {
+    // TX_START = 32 + MAX_INPUTS * (32 + 8) = 32 + 8 * (40) = 352
+    // inputs   = TX_START + 12 words = 352 + 96             = 448
+
+    let input_start = asm(r1, r2: n) {
+        xis r1 r2;
+        r1: u64
+    };
+
+    let input_length = asm(r1, r2: n) {
+        xil r1 r2;
+        r1: u64
+    };
+
+    // Inputs begin at the 12th word in a TransactionScript
+    asm(buffer, start: input_start, length: input_length, inputs_ptr: 448) {
+        move buffer sp;
+        mcp buffer input_start input_length;
+        buffer: u32
+    }
```

Comment:
This is a bit dangerous without a `CFEI`.  The compiler assumes `SP` is pointing to the end of the stack at all times.